### PR TITLE
Add Cocktail Sort

### PR DESCRIPTION
**Pull request by migueldalberto**
The cocktail sort (also known as cocktail shaker sort, bidirectional bubble sort, shaker sort, ripple sort, shuffle sort, or shuttle sort) is a variant of the bubble sort algorithm. It extends the bubble sort by operating in both directions (from array's beginning to its end, and from array's end to its beginning).

I based the implementation on the [Wikipedia page about Cocktail Sort](https://en.wikipedia.org/wiki/Cocktail_shaker_sort).

Also added this sort on the `sort_tests.go`

### CODE REVIEW

**Review by raklaptudirm — COMMENTED** [review]
Add some inline comments explaining the algorithm as far as possible.

File: sort/cocktailsort.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// Cocktail sort is a variation of bubble sort, operating in two directions (beginning to end, end to beginning)
+func Cocktail[T constraints.Integer](arr []T) []T {
+	if len(arr) == 0 { // ignore 0 length arrays
+		return arr
+	}
+
+	swapped := false // true if swapped two or more elements in the last loop
+	// if it loops through the array without swapping, the array is sorted
+
+	for ok := true; ok; ok = swapped { // exits when swapped == false
```
Comment (raklaptudirm) [inline sort/cocktailsort.go:17]: Isn't this whole mess just equivalent to:
```go
for swapped {
```
Not sure what I am missing.

I guess that's to ensure the loop doesn't break in the first iteration? That can be fixed by simply setting the initial value of `swapped` to `true`.

File: sort/cocktailsort.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+		for i := 0; i < len(arr)-1; i++ { // first loop, from array's beginning to end
```
Comment (raklaptudirm) [inline sort/cocktailsort.go:19]: You probably need to set swapped to false before this loop.

**Review by raklaptudirm — CHANGES_REQUESTED** [review]
I would suggest adding `start` and `end` variables to make the sort more efficient.

File: sort/cocktailsort.go (hunk 2 above)
Comment (migueldalberto) [reply sort/cocktailsort.go:19]: I thought about doing that but preferred the way I did instead. Maybe a more readable solution is to just put in a `for {}` with another if exit at the end.

**Comment by migueldalberto** [comment]
> I would suggest adding `start` and `end` variables to make the sort more efficient.

Added

File: sort/cocktailsort.go (hunk 2 above)
Comment (raklaptudirm) [reply sort/cocktailsort.go]: This code:
https://github.com/TheAlgorithms/Go/blob/4c573a955c61932ec1513d90f92f5177636fa3cb/sort/cocktailsort.go#L52
Is equivalent to:
```go
for swapped {
```
It just uses more code and locs, so I am not sure why you think its better.

Comment (migueldalberto) [reply sort/cocktailsort.go]: Done.

**Review by raklaptudirm — APPROVED** [review]

File: sort/cocktailsort.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
+// Implementation of Cocktail sorting
+// reference: https://en.wikipedia.org/wiki/Cocktail_shaker_sort
+
+package sort
+
+import "github.com/TheAlgorithms/Go/constraints"
+
+// Cocktail sort is a variation of bubble sort, operating in two directions (beginning to end, end to beginning)
+func Cocktail[T constraints.Integer](arr []T) []T {
```
Comment (tjgurwara99) [inline sort/cocktailsort.go:9]: From my understanding, isn't this algorithm applicable to all `Ordered` types?

Comment (migueldalberto) [reply sort/cocktailsort.go]: Right.

**Review by tjgurwara99 — APPROVED** [review]

**Review by raklaptudirm — APPROVED** [review]

**Pull request merged**